crates/ide-assists/src/handlers/convert_nested_function_to_closure.rs

```diff
@@ -0,0 +1,186 @@
+use ide_db::assists::{AssistId, AssistKind};
+use syntax::ast::{self, HasGenericParams, HasName};
+use syntax::{AstNode, SyntaxKind};
+
+use crate::assist_context::{AssistContext, Assists};
+
+// Assist: convert_nested_function_to_closure
+//
+// Converts a function that is defined within the body of another function into a closure.
+//
+// ```
+// fn main() {
+//     fn fo$0o(label: &str, number: u64) {
+//         println!("{}: {}", label, number);
+//     }
+//
+//     foo("Bar", 100);
+// }
+// ```
+// ->
+// ```
+// fn main() {
+//     let foo = |label: &str, number: u64| {
+//         println!("{}: {}", label, number);
+//     };
+//
+//     foo("Bar", 100);
+// }
+// ```
+pub(crate) fn convert_nested_function_to_closure(
+    acc: &mut Assists,
+    ctx: &AssistContext<'_>,
+) -> Option<()> {
+    let name = ctx.find_node_at_offset::<ast::Name>()?;
+    let function = name.syntax().parent().and_then(ast::Fn::cast)?;
+
+    if !is_nested_function(&function) || is_generic(&function) {
+        return None;
+    }
+
+    let target = function.syntax().text_range();
+    let body = function.body()?;
+    let name = function.name()?;
+    let params = function.param_list()?;
+
+    acc.add(
+        AssistId("convert_nested_function_to_closure", AssistKind::RefactorRewrite),
+        "Convert nested function to closure",
+        target,
+        |edit| {
+            let has_semicolon = has_semicolon(&function);
+            let params_text = params.syntax().text().to_string();
+            let params_text_trimmed =
+                params_text.strip_prefix("(").and_then(|p| p.strip_suffix(")"));
+
+            if let Some(closure_params) = params_text_trimmed {
+                let body = body.to_string();
+                let body = if has_semicolon { body } else { format!("{};", body) };
+                edit.replace(target, format!("let {} = |{}| {}", name, closure_params, body));
+            }
```

Comment:
When `params_text_trimmed` is `None` IDE will show the assist without any edit attached. Can you move this condition outside `acc.add()`?